lib/ash/resource/validation/compare.ex

```diff
@@ -34,41 +34,30 @@ defmodule Ash.Resource.Validation.Compare do
 
     case value do
       {:ok, value} ->
-        Enum.reduce(
-          Keyword.take(opts, [
-            :greater_than,
-            :less_than,
-            :greater_than_or_equal_to,
-            :less_than_or_equal_to
-          ]),
-          :ok,
-          fn validation, _ ->
-            case validation do
-              {:greater_than, attribute} ->
-                if Comp.greater_than?(value, attribute_value(changeset, attribute)),
-                  do: :ok,
-                  else: invalid_attribute_error(opts, value)
-
-              {:greater_than_or_equal_to, attribute} ->
-                if Comp.greater_or_equal?(value, attribute_value(changeset, attribute)),
-                  do: :ok,
-                  else: invalid_attribute_error(opts, value)
-
-              {:less_than, attribute} ->
-                if Comp.less_than?(value, attribute_value(changeset, attribute)),
-                  do: :ok,
-                  else: invalid_attribute_error(opts, value)
-
-              {:less_than_or_equal_to, attribute} ->
-                if Comp.less_or_equal?(value, attribute_value(changeset, attribute)),
-                  do: :ok,
-                  else: invalid_attribute_error(opts, value)
-
-              true ->
-                :ok
-            end
-          end
-        )
+        opts
+        |> Keyword.take([
+          :greater_than,
+          :less_than,
+          :greater_than_or_equal_to,
+          :less_than_or_equal_to
+        ])
+        |> Enum.find_value(fn
```

Comment:
So the original point of the reduce (although not working) was to gather up multiple errors. So for this, lets turn this into a `flat_map`, where the `:ok` case returns `[]` and the error returns `[the_error]`, and then check do something like:

```elixir
|> case do
  [] -> :ok
  errors -> {:error, errors}
end
```

Reply:
Gotcha. Will make that change tomorrow 👍

Reply:
Actually, thinking about this a bit more - how could a `compare` check (in a `numericality` validation) generate more than one error on a field? All of the options seem pretty mutually exclusive, unless you have odd conditions like `[greater_than: 5, greater_than_or_equal_to: 10]` and you pass in a number that fails both checks.

The only case I can think of is if we wanted to validate that a number was _outside_ a given range, eg x < 5 || x > 10, but that would have to be implemented with negating a standard numericality validation, wouldn't it?

The `message` function also prints out the full conditions (eg. "must be greater than 1 and must be less than 100") for the error, regardless of which part failed.

Reply:
You could be validating a range, like greater than 1 and less than 10.

Reply:
But the fact that the message contains all the info does suggest that we could just add the one error.

Reply:
Even if you are validating a range, there's only going to be one error - you can't fail both checks at once.